File: eunji/0x08/BOJ_3986.py
from collections import deque
import sys

def check(func):
    D = deque([])
    A_flag = 0; B_flag = 0
    for i in func:
        if i == 'A':
            if (A_flag > 0 and D[-1] == 'A'):
                D.pop()
                A_flag -= 1
            else:
                D.append('A')
                A_flag += 1
        elif i == 'B':
            if (B_flag > 0 and D[-1] == 'B'):
                D.pop()
                B_flag -= 1
            else:
                D.append('B')
                B_flag += 1
                
    if len(D) == 0: return 1
    else: return 0
    

N = int(sys.stdin.readline())
ans = 0
for i in range(N):
    func = sys.stdin.readline()[:-1]
    ans += check(func)
print(ans)

File: eunji/0x08_stackApp/BOJ_3986.cpp
#include <iostream>
#include <deque>
using namespace std;


int check(string func){
    deque<char> D;
    int A_flag = 0;
    int B_flag = 0;
    for (auto c : func){
        if (c == 'A'){
            if (A_flag > 0 && D.back() == 'A') {
                D.pop_back();
                A_flag --;
            }
            else {
                D.push_back('A');
                A_flag ++;
            }
        }
        else if (c == 'B'){
            if (B_flag > 0 && D.back() == 'B') {
                D.pop_back();
                B_flag --;
            }
            else {
                D.push_back('B');
                B_flag ++;
            }
        }
    }
    if (D.size() == 0) return 1;
    else return 0;
}

int main(void){
    cin.tie(0);
    cin.sync_with_stdio(0);
    int N;
    cin >> N;
    int ans = 0;
    for (int i=0; i<N; i++){
        string func;
        cin >> func;
        ans += check(func);
    }
    cout << ans;
}